src/opensles/AudioStreamBuffered.cpp

```diff
@@ -25,57 +25,187 @@ namespace oboe {
+            result = mFifoBuffer->read(data, framesLeft);
+        } else {
+            result = mFifoBuffer->write(data, framesLeft);
+        }
         if (result > 0) {
-            source += mFifoBuffer->convertFramesToBytes(result);
-            incrementFramesWritten(result);
+            data += mFifoBuffer->convertFramesToBytes(result);
             framesLeft -= result;
         }
-        if (framesLeft > 0 && result >= 0) {
-            int64_t wakeTimeNanos =  mFifoBuffer->getNextReadTime(getSampleRate());
-            // TODO use timeoutNanoseconds
-            AudioClock::sleepUntilNanoTime(wakeTimeNanos);
+
+        // If we need more data then sleep and try again.
+        if (framesLeft > 0 && result >= 0 && timeoutNanoseconds > 0) {
+            int64_t timeNow = AudioClock::getNanoseconds();
+            if (timeNow >= timeToQuit) {
+                LOGE("AudioStreamBuffered::%s(): TIMEOUT", __func__);
+                repeat = false; // TIMEOUT
+            } else {
+                // Figure out how long to sleep.
+                int64_t sleepForNanos;
+                int64_t wakeTimeNanos = predictNextCallbackTime();
+                if (wakeTimeNanos <= 0) {
+                    // No estimate available. Sleep for one burst.
+                    sleepForNanos = (getFramesPerBurst() * kNanosPerSecond) / getSampleRate();
+                } else {
+                    // Don't sleep past timeout.
+                    if (wakeTimeNanos > timeToQuit) {
+                        wakeTimeNanos = timeToQuit;
+                    }
+                    sleepForNanos = wakeTimeNanos - timeNow;
+                    // Avoid rapid loop with no sleep.
+                    const int64_t minSleepTime = kNanosPerMillisecond; // arbitrary
+                    if (sleepForNanos < minSleepTime) {
+                        sleepForNanos = minSleepTime;
+                    }
+                }
+
+                AudioClock::sleepForNanos(sleepForNanos);
+            }
+
+        } else {
+            repeat = false;
         }
+    } while(repeat);
+
+    return result < 0 ? result : (numFrames - framesLeft);
+}
+
+// Write to the FIFO so the callback can read from it.
+int32_t AudioStreamBuffered::write(const void *buffer,
+                                   int32_t numFrames,
+                                   int64_t timeoutNanoseconds) {
+    if (getDirection() == Direction::Input) {
+        return (int32_t) Result::ErrorUnavailable;  // TODO review, better error code?
+    }
+    return transfer((void *) buffer, numFrames, timeoutNanoseconds);
+}
+
+// Read data from the FIFO that was written by the callback.
+int32_t AudioStreamBuffered::read(void *buffer,
+                                  int32_t numFrames,
+                                  int64_t timeoutNanoseconds) {
+    if (getDirection() == Direction::Output) {
+        return (int32_t) Result::ErrorUnavailable; // TODO review, better error code?
```

Comment:
ErrorInvalidState?

Reply:
or maybe ErrorUnsupported

Reply:
I would categorize writing to an input stream or reading from an output stream as invalid operations, so ErrorInvalidOperation?

Reply:
I like InvalidOperation. Currently all oboe errors are defined as AAUDIO_ERROR_*. We could add INVALID_OPERATION to aaudio and try keep the numbers in sync.